Better flat stylebox with rounded corners

toger5 opened the pull request:
This is still kind of WIP
**NOW it's almost done though, it can be used but some features would still be nice to get added, check issues at the end of this post**

## what is supported for now:

UPDATED:
```
//Color
	void set_bg_color(const Color &p_color);
	Color get_bg_color() const;

	//Border Color
	void set_border_color_all(const Color &p_color);
	void set_border_color(Margin p_border, const Color &p_color);
	Color get_border_color(Margin p_border) const;

	void set_light_color(const Color &p_color);
	Color get_light_color() const;

	void set_dark_color(const Color &p_color);
	Color get_dark_color() const;

	//BORDER
	//width
	void set_border_width_all(int p_size);
	int get_min_border_width() const;

	void set_border_width(Margin p_margin, int p_size);
	int get_border_width(Margin p_margin) const;

	//blend
	void set_border_blend(bool p_blend);
	bool get_border_blend() const;

	//CORNER_RADIUS
	void set_corner_radius_all(int radius);
	void set_corner_radius_individual(const int radius_top_left, const int radius_top_right, const int radius_botton_right, const int radius_bottom_left);
	int get_min_corner_radius() const;

	void set_corner_radius(Corner p_corner, const int radius);
	int get_corner_radius(Corner p_corner) const;

	//EXPANDS
	void set_expand_margin_size(Margin p_expand_margin, float p_size);
	float get_expand_margin_size(Margin p_expand_margin) const;

	//FILLED
	void set_filled(bool p_draw);
	bool is_filled() const;

	//SHADOW
	void set_shadow_color(const Color &p_color);
	Color get_shadow_color() const;

	void set_shadow_size(const int &p_size);
	int get_shadow_size() const;
```

most of this is relatively easy to understand.
basically allows for rounded corners and different color for each corner.

## how it works:

old version:
~~I draw rounded rectangles by first adding circles in the corners with the right offset and the right radius. than I draw four rectangles for each edge. (need for because it supports different radius for each corner).~~

I draw rounded rings by creating two arrays of certs which have a defined gap between them and fill them with polygons + set the color according to settings (if border_blend inner and outer arrays have different colors)
I also use that function to draw the shadow. with outer color = transparent
The aa will be done by creating another outer ring around the border and the center which will fade to transparent too.
The infill has other indexing logic but I also generate a vertex array in the shape of a rounded ring. just one though and than fill it with triangles from right to left.

## issues:
 - [x] empty centers are not possible with rounded corners the circles are no than inside the circle:
<img width="1680" alt="screen shot 2017-05-24 at 13 54 41" src="https://cloud.githubusercontent.com/assets/16718859/26425705/ae471a78-408a-11e7-818d-3cea2a22b899.png"> this could be fixed if I have some kind of masking feature, or if I draw 90 degree arc's instead of circles for style boxes without center... (second might be easiest)

 - [x] the history needs some cleanup
- [x] There is an issue with the circle drawing. it seems like only every second (or random...) circle is drawn. It works when I just do draw 2 circles on top of each other... which I do right not.

 - [x] the polygon drawing is broken... so I copied code from this pr: #8425 

 
 - [x] there are still some comments which should get removed
- [x] border color for rounded is still off
 - [x] wrong type for GDScript and Inspector (float should be color) 
 - [x] reimplementing everything with polygons
 - [x] make blend and fill work with polygons
 - [x] shadows
 - [x] different border colors with polygons
 - [x] different border width with polygons
 - [x] expand margin
 - [x] rename: set_all_border_width -> set_border_width_all
 - [x] rename set_all_corner_radius -> set_corner_radius_all
 - [x] rename set_all_border_color -> set_border_color_all
- [x] decide to only allow half_height corner radius and border or improove the limit calculation (also take the corner radius + the border into account) (needs to be done I NEED FEEDBACK WHAT BEHAVIOUR IS WANTED)
- [x] antiAliasing
- [x] anti_aliasing sizing (*grow(-1)*) (@n-pigeon)
- [x] corner_segments as a property ( @n-pigeon )
- [x] corner bug which only shows 100 * (corner_detail - 1) / corner_detail
- [x] infill not fully drawn (missing last try)
- [x] additional border is not supported anymore. need to use expand margin + border width. Change that in the default theme! 
- [x] **individual corner and border color** `CANCELD FOR FIRST VERSION`
- [x] color gradient (**needed?**) `CANCELD FOR FIRST VERSION`
- [x] added documentation.

toger5 commented:
I did another update.
updated the todoList.
Here are some screenshots of how it looks (just examples, I'm not claiming this makes the default theme look better) I think the flat design is kind of cool but some rounded corners might make it look better...
I think the biggest use case is in custom themes for games and other themes for the editor.

current version.
<img width="1680" alt="screen shot 2017-05-24 at 18 33 16" src="https://cloud.githubusercontent.com/assets/16718859/26432419/8beed4f4-40b0-11e7-9ecb-93dab6a5c9f0.png">

applying rounded corners to everything (older version -> broken ...)

<img width="1680" alt="screen shot 2017-05-24 at 05 44 52" src="https://cloud.githubusercontent.com/assets/16718859/26432471/e0e711ce-40b0-11e7-933a-d2a812bcc3ae.png">


how top left right bottom colors and border blend behave with rounded corners: 

<img width="243" alt="screen shot 2017-05-24 at 18 44 30" src="https://cloud.githubusercontent.com/assets/16718859/26432505/0f6a7338-40b1-11e7-80f3-0a64dfb05bc3.png">

toger5 commented:
@volzhs since you are very familiar with style boxes could you do some testing?
I wil try to keep this branch up to date, so that you can test with current master.

I can't change line width so I'm using multiple lines. I think performance wise the whole rounded style box thing is a desaster.. if someone could give some tips/fix some gels 3 functions to make it work would be highly appreciated.

toger5 commented:
Fixed a couple of more things performance wise and added a couple of more features.

n-pigeon commented:
Yo, found few problems :)

Image worth 1000 words:
![selection_101](https://cloud.githubusercontent.com/assets/6152134/26520851/fd72bb3c-42da-11e7-9fcc-3cdb91832551.png)

Things that may be unnoticed.
Props for borders color are `float` instead of `color`. Fill and border don't match resulting transparent pixels in between.

There are also problems when you modulate color to change Opacity. Original StyleBoxFlat also had problem with it ^_^'

![selection_102](https://cloud.githubusercontent.com/assets/6152134/26520874/929ff65c-42db-11e7-8e0e-d7dd468df853.png)

toger5 commented:
Here is a branch with some theme changes using rounded style boxes:
https://github.com/toger5/godot/tree/styleBoxAndThemeChanges
just clone and compile

toger5 commented:
@n-pigeon Thank you! 
the color, Int error should be fixed.

toger5 commented:
@n-pigeon 
the transparency bug say is not as easy to fix. I would need to build a vertex shape for the whole inner part so it would look good. (currently I'm drawing circles for the corners and rects for the center)

I also draw 10 lines max, reason is, that you can get it really buggy and slow if you allow more...

I think at the end I will have to go for full polygons anyways... damn that will be a little more work ;)

or I limit border radius to 10.. or 20 might still run fine.

too big of a limitation?

toger5 commented:
rebased. on master

@akien-mga should I squash all the commits? or separate them in 1-3?
Since I rely on #8425, should I wait until this one is fixed anyways?
(of course this branch needs to get approved by u guys and tested properly but is still can start with squashing)

n-pigeon commented:
Full polygon will be probably needed sooner or later :) If you will need some help with this stuff I would gladly help.
Biggest issue will be AA. I have no idea how to make it :D

Zylann commented:
@toger5 did you see my code comments?

toger5 commented:
@Zylann sorry I did not. and I think the rebase removed them. let me check my mail if I still can get them.

Zylann commented:
Looks like I forgot to click the "Submit review" button. I'm not used to this yet^^

toger5 commented on scene/resources/style_box.cpp:
```diff
@@ -393,6 +582,19 @@ void StyleBoxFlat::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("get_corner_radius_TL"), &StyleBoxFlat::get_corner_radius_TL);
+	ClassDB::bind_method(D_METHOD("get_corner_radius_TR"), &StyleBoxFlat::get_corner_radius_TR);
+	ClassDB::bind_method(D_METHOD("get_corner_radius_BL"), &StyleBoxFlat::get_corner_radius_BL);
+	ClassDB::bind_method(D_METHOD("get_corner_radius_BR"), &StyleBoxFlat::get_corner_radius_BR);
```
that makes a lot of sense since it also looks super wired in the inspector

toger5 replied:
changed

toger5 commented on scene/resources/style_box.cpp:
```diff
@@ -316,67 +333,239 @@ bool StyleBoxFlat::get_border_blend() const {
+		if (filled) {
+			//the circle has to be drawn twice, otherwise the circles for the corners are not visible
+			//super buggy ;)
+			vs->canvas_item_add_circle(p_canvas_item, corners[i], (float)corner_radius[i], col);
```
is one draw now

toger5 replied:
it is one draw now

toger5 replied:
filled arc is a really good idea!! also fixes some of the artefacts when drawing with transparency

toger5 commented on scene/resources/style_box.cpp:
```diff
@@ -316,67 +333,239 @@ bool StyleBoxFlat::get_border_blend() const {
+			rad = 0;
+		inner_corner_radius[i] = rad;
+	}
+	int *p = &inner_corner_radius[0];
```
```
inline PoolIntArray get_offset_corner_radius(int offset, PoolIntArray corner_radius) {
	PoolIntArray inner_corner_radius = PoolIntArray();
	for (int i = 0; i < 4; i++) {
		int rad = corner_radius[i] - offset;
		if (rad < 0)
			rad = 0;
		inner_corner_radius.append(rad);
	}
	return inner_corner_radius;
}```
is what I changed it to

Zylann replied:
well it's not that good either, you are allocating an array everytime. I don't understand why you created this function in the first place tbh, I feel like you don't really need it

toger5 replied:
I need to recalculate the corner radius in a couple of situations: drawing inner part of the box + drawing the border lines...
Should I allocate the memory inside the draw_rounded_rect function and just pass a ref to the function which modifies it?

Zylann replied:
My point is you probably don't need to allocate any memory^^

toger5 replied:
thank you for the explanation on read write objects. At the beginning I thought that would be the only way to use poolArray. than I found out you can use []... seemed so much simpler ;) 
But now I know...

toger5 commented:
I was thinking that I might switch to a full polygon implementation. This results in better performance, cleaner code, better looks and also the option for border thickness.

border thickness would be great:
currently there is additional border size which was implemented by @volzhs, it works by just adding a rectangle to the desired border. It solves a similar problem than expand margin for style box texture. But expand margin just sizes the stylebox outside the boundaries.
For consistency I would like them to both have the same behaviour. With rounded corners thats needed anyways. And the implementation of border thickness would allow to achieve both with just expand margin.

n-pigeon commented:
I was thinking about ability to control thickness for all sides not just one global.

toger5 commented:
Yea thats also what  I was proposing
(bottom left in the sketch: border width per border)
It is a little chellanging how to decide on on corner radius and shape for inner and outer one without flooding the user with options. But I figured something out, just need to implement it.

I have decided now to reimplement the whole thing with polygons.

Hope ill find time soon

Zylann commented:
What do you mean with polygons? Also I implemented a line mesh builder in scene/2d/ if you need one ;)

toger5 commented:
@Zylann Currently I'm using draw rect, draw circle, draw line (of course the circle is using polygons too at the end) For the border I literally draw lines around the center... (got inspired by the super Hacky previous implementation of style box rendering: for the blend border multiple boxes were just drawn on top of each other with different sizes and colors)
I will calculate the vert array and color+ indices inside the draw function and than just do one draw call.
that is much easier to read, better performance, and less hacky.
I will checkout the line_mesh builder, but I already did the pseudocode for the whole poly shape with indexing+ color, infill... Maybe it's still easier with your line mesh builder.
With my implementation I can also do really easy shadows and fake anti aliasing. (anti aliasing by adding outer ring of verts with distance one to the box itself and than blend the colors to transparent)

volzhs commented:
@toger5 I like that approach.

toger5 commented:
Reimplemented the whole thing with polygons!
not feature complete yet:
still need shadows,
borders without out blending (currently everything is bending because I can't give individual indices colors).
The expand margin I proposed,
multiple border colors,
and some renaming of variables/bindings

toger5 commented:
@volzhs I the branch with the poly implementation is uploaded. Ready for testing.

toger5 commented:
This is the current naming for styleBoxFlat:

```
//Color
	void set_bg_color(const Color &p_color);
	Color get_bg_color() const;

	void set_all_border_color(const Color &p_color);

	void set_light_color(const Color &p_color);
	Color get_light_color() const;

	void set_dark_color(const Color &p_color);
	Color get_dark_color() const;

	//Border Color
	void set_border_color(const Color &p_color, const Margin &p_border);
	Color get_border_color(const Margin &p_border) const;

	void set_border_color_left(const Color &p_color);
	void set_border_color_top(const Color &p_color);
	void set_border_color_right(const Color &p_color);
	void set_border_color_bottom(const Color &p_color);
	Color get_border_color_left();
	Color get_border_color_top();
	Color get_border_color_right();
	Color get_border_color_bottom();

	//BORDER
	//blend
	void set_border_blend(bool p_blend);
	bool get_border_blend() const;
	
	//width
	void set_all_border_width(int p_size);
	int get_min_border_width() const;

	void set_border_width(Margin p_margin, int p_size);
	int get_border_width(Margin p_margin) const;

	void set_border_widht_left(int p_size);
	void set_border_widht_top(int p_size);
	void set_border_widht_right(int p_size);
	void set_border_widht_bottom(int p_size);
	int get_border_widht_left();
	int get_border_widht_top();
	int get_border_widht_right();
	int get_border_widht_bottom();
	
	//CORNER_RADIUS
	//set
	void set_all_corner_radius(int radius);
	PoolIntArray get_corner_radius() const;

	void set_corner_radius(const int radius_top_left, const int radius_top_right, const int radius_botton_right, const int radius_bottom_left);
	int get_min_corner_radius() const;

	void set_corner_radius_TL(int radius);
	void set_corner_radius_TR(int radius);
	void set_corner_radius_BR(int radius);
	void set_corner_radius_BL(int radius);
	int get_corner_radius_TL() const;
	int get_corner_radius_TR() const;
	int get_corner_radius_BR() const;
	int get_corner_radius_BL() const;

	void set_filled(bool p_draw);
	bool is_filled() const;
```

Thoughts?
what should I do with set_all...
should I just do overloaded functions for set_corner_radius ... one with one property and another one with margin + value and another one with four properties for all four corners, borders?

n-pigeon commented:
I'm not sure about `_radius_TL` it should be at least `_radious_tl`.

Maybe `_radius_t_l` then it will be displayed in properties as: `Radius T L`.

toger5 commented:
I did another big renaming. Can't get it to compile right now though: Some issue with bindings to ClassDB...
I got rid of all the `_tl`, `_left` functions. only leaving `set_(margin)` `get_(margin)` this is much better since the bidnigs allow to still use those functions for properties.
so in c++ you do: `box.set_border_width(MARGIN_LEFT, 20);`
in gdScript: `box.border_width_left = 20`
this tremendously reduces set, get functions in the style box and still allows for the same usability in gdScript

toger5 commented:
I continued with a couple of features.
And even more importatn: renamed everything ;) and brought it into proper structure.

##new (and probably final) naming for the new styleboxes.
(I might add one feature for different corner color settings...)
<img width="452" alt="screen shot 2017-05-29 at 17 48 04" src="https://cloud.githubusercontent.com/assets/16718859/26564556/3d887da2-4497-11e7-9a2d-669d0bcb3848.png">

This is what you already can do with it:

<img width="151" alt="screen shot 2017-05-29 at 17 47 31" src="https://cloud.githubusercontent.com/assets/16718859/26564555/3d66f7ae-4497-11e7-9161-c457fc95ada6.png">

Geequlim commented:
Is shadows and color gradient in the plan？

toger5 commented:
shadows are in plan. will do them now.
What do you mean with color gradient? There are gradients for the border (border_blend).
but I guess your are talking about infill gradients. It would be relatively easy to make a top and bottom bg_color with gradient. left right is also possible but harder...
the gradient actually would look pretty good with poly's. don't know what you would use it for.

@Geequlim how would you make the settings? just a top + bottom color or a bool setting and top and bottom color. Than you don't have to change both colors each time you want to change the bg.

Geequlim commented:
@toger5 I think it is enough with `top to bottom` and `left-to-right` color gradient for most use case.

toger5 commented:
@Geequlim what settings do you think would work best for that? (what variables should be exposed to achieve different gradients?) is there a gradient Variant ?

toger5 commented:
maybe like this:
```
Color bg_color
Color second_bg_color
Enum gradient {NONE, TOP_BOTTOM, LEFT_RIGHT}
```
than second_bg_color can be left untouched and only takes effect when a gradient is choosen `(gradient != NONE)`

toger5 commented:
I just did another patch. the whole style box thing now is in a pretty good change. It still is possible to create buggy style boxes but that was possible before also. So the style box creator needs to know what he does...
here is a small example what you can do: (might play around a little bit now so there will be more examples soon)
<img width="135" alt="screen shot 2017-05-30 at 00 04 31" src="https://cloud.githubusercontent.com/assets/16718859/26572477/f015bbe2-44ce-11e7-9d64-409f45ba7b5a.png">

toger5 commented:
This is the new sorting of the properties in the inspector:
<img width="468" alt="screen shot 2017-05-30 at 01 06 34" src="https://cloud.githubusercontent.com/assets/16718859/26573892/4508e494-44d4-11e7-9b68-13a6686ee6be.png">

feedback as always highly appreciated...

toger5 commented:
@akien-mga do you think we can already merge (after some testing) or should I add the missing features mentioned in the Issue section of the first post?

If so (merge) I will send you a `rebase -i` proposal which commits I would squash. Or do you want everything in one commit? (I think I would separate changes like: `adding Enum to math.h`.)

n-pigeon commented:
Good job!

I would ask someone maybe @reduz if there would be a way to make AA for round stylebox less hacky way :)

Zylann commented:
Can they be baked into textures so that drawing them over and over with AA and stuff is less expensive?

toger5 commented:
@Zylann right now one style box is: 4 * 8 * 2 + 16 * 2 (infill) + (optional shadow) 4 * 8 * 2 + and aa would be another 2 * (4 * 8 * 2) verts. per style box

n-pigeon commented:
It should be that heavy.

Looking for other software Blender renders it's UI dynamically in RT with OGL and they have huge amount of widgets on screen.

toger5 commented:
@n-pigeon so you mean I only draw two triangles and than do one shader?

good idea let me think about it and do some tests.

Zylann commented:
@toger5 gradients also work by giving colors to vertices, and the default pixel shader will interpolate them (unless you do that already?)

toger5 commented:
@Zylann thats what I'm doing right now. for aa and for gradients. but if I could write a shader for everything would be even better... I think

Zylann commented:
I don't see how a shader would be better, is there a very special case where vertex color doesn't cut it? (swapping shaders has a cost too)

n-pigeon commented:
What I meant was to draw the shape with polygons and create gradients in this poly with shader :)

toger5 commented:
for the gradients (with a shader) I would need to use position anyways so I need to do the circle math for it.. than I also could just do everything with shaders.
A shader would be better by allowing for fewer vertices. but I think the performance is still pretty good. Most games don't have too many ui elements and the editor needs a lot of really simple styles. I payed a lot of attention that those simpler style boxes are not getting harder to draw...

toger5 commented:
@n-pigeon anti aliasing for 2d shapes and lines is not too easy. there is not glEanable(ANTI_ALIASING). at least I don't know about it. so you have to draw some kind of blur around it. and I think the difference between a shader and vertices is not that big since It also takes some computing to find out if the pixel is at the border of the shape...

toger5 commented:
FINALLY got antiAliasing done!!!

standart...
<img width="137" alt="screen shot 2017-06-01 at 01 34 37" src="https://cloud.githubusercontent.com/assets/16718859/26671583/decd163a-466a-11e7-9cfa-7d1b8bd10db4.png">
<img width="131" alt="screen shot 2017-06-01 at 01 34 41" src="https://cloud.githubusercontent.com/assets/16718859/26671584/dece17f6-466a-11e7-9e7e-d4e177d8ca55.png">

a more advanced shape possible with the new styleboxes
<img width="144" alt="screen shot 2017-06-01 at 01 36 53" src="https://cloud.githubusercontent.com/assets/16718859/26671585/ded07ffa-466a-11e7-8b93-3486248f3917.png">
<img width="135" alt="screen shot 2017-06-01 at 01 36 58" src="https://cloud.githubusercontent.com/assets/16718859/26671582/decb9800-466a-11e7-984b-57b3261f3b78.png">

hope it is obvious which one is which...

toger5 commented:
I uploaded a small (in terms of resolution... Damn it) youtube video where I demonstrate a button using the new styleBoxFlat.
https://youtu.be/SYKPARDR1J0

Zylann commented on scene/resources/style_box.cpp:
```diff
@@ -776,41 +797,38 @@ void StyleBoxFlat::draw(RID p_canvas_item, const Rect2 &p_rect) const {
-		Color t = Color(0, 0, 0, 0);
-		Color transparent[4] = { t, t, t, t };
 		Color shadow_col_arr[4] = { shadow_color, shadow_color, shadow_color, shadow_color };
 		drawRing(verts, indices, colors, style_rect, adapted_corner,
```
drawRing snake case plz :3

toger5 replied:
damn of course...

n-pigeon commented:
Woo nice!
Oh, oh idea! Maybe additional parameter to control the amount of vertices n the corner? If somebody want's more smooth corner he can up the amount and it will allow for some cool bevel like effects in lower vert numbers :D

toger5 commented:
@n-pigeon good idea! I add it to the to do

nhydock commented:
Are all properties tweenable?  I'd like to use this with animation player so I can make transitions that are a bit smoother, like I can with CSS.  If that's possible then I'm super excited for this.

toger5 commented:
@nhydock I'm pretty sure they are: tween just changes the value over time as far as I know, since the style box gets redrawn after each change it should animate just fine.

Zylann commented:
To make this work you need several things:

1) Make the style of the widget you want to animate has a unique StyleBoxFlat resource, so that your animation won't modify every widget using this style

2) If StyleBoxFlat has no property, you can create a script that exposes one and modifies the underlying StyleBoxFlat

3) Call update() on the widget that you want to animate while the animation is playing

Zylann commented:
Also, given all these nifty options, should this be renamed? StyleBoxFlat is no longer really flat^^

toger5 commented:
@Zylann good point.
what are you thinking about.
 - StyleBoxGeneric
 - StyleBoxBasic
 - StyleBoxDefault
 - StyleBoxGenerated
 - StyleBoxSimple

toger5 commented:
- StyleBoxProcedual

toger5 commented:
@n-pigeon I implemented the variable detail lvl. It is not as easy than expected though...
BUT it also helped me find another issue. So I have to work on it anyways:
currently not the whole corner is drawn. a corner is PI/2 rad, but I don't draw the last polys. if the detail is 8 I only draw 7 parts -> from 0 to PI/2 * 7/8 which is not the whole corner... as a consequence the top and bottom line are not completely horizontal. this leads to a one pixel offset:
<img width="98" alt="screen shot 2017-06-01 at 23 40 09" src="https://cloud.githubusercontent.com/assets/16718859/26713896/af60bbfe-4723-11e7-8ab8-8dcfcbb99743.png">
I never could figure out what it was but now I know...

I still don't know about your proposal:
It might be really hard that the border thickness is correct making a beveled rectangle with a flat bevel. (detail 1)

![styleboc neede chagnes](https://cloud.githubusercontent.com/assets/16718859/26713756/16e815fc-4723-11e7-8a98-0bdcaf970050.png)

As you can see in the screenshot, the border would be thinner on the edge than on the border... with higher detail that is not visible. but low detail makes it pretty obvious...
I would need to change the way the inner verts are calculated. I don't know if this is worth it... I still will keep the option and for smaller corner radius it will work just fine... and fore some situations it might be the needed behaviour.
If I want detail 1 (beveled) corners to look the same with than I need to adapt. In other word write a specific rule for how it should behave with corner detail 1 if I want the border to be the same width.
I usually try to avoid special rules as much as I can... Maybe I find a way to make it always look good. otherwise we can decide to keep the setting although the behaviour is not the best

toger5 commented:
Fixed the corner bug:
before:
<img width="131" alt="screen shot 2017-06-02 at 10 35 11" src="https://cloud.githubusercontent.com/assets/16718859/26739175/afed3266-4785-11e7-97e1-1127531e51d5.png">
after:
<img width="127" alt="screen shot 2017-06-02 at 10 37 37" src="https://cloud.githubusercontent.com/assets/16718859/26739594/3aa89b1a-4787-11e7-8218-3a1267d7b6b6.png">

toger5 commented:
some more spam...
<img width="129" alt="screen shot 2017-06-02 at 10 41 12" src="https://cloud.githubusercontent.com/assets/16718859/26746639/b89504aa-47a5-11e7-86b6-7b9d0b929230.png">
<img width="135" alt="screen shot 2017-06-02 at 15 07 21" src="https://cloud.githubusercontent.com/assets/16718859/26746644/b89d2e3c-47a5-11e7-8471-acfdb5c57b69.png">
<img width="142" alt="screen shot 2017-06-02 at 15 09 00" src="https://cloud.githubusercontent.com/assets/16718859/26746643/b89bec84-47a5-11e7-8b23-e4114667f920.png">
<img width="146" alt="screen shot 2017-06-02 at 15 09 15" src="https://cloud.githubusercontent.com/assets/16718859/26746642/b89b1f16-47a5-11e7-898d-6e0961ba33e4.png">
<img width="150" alt="screen shot 2017-06-02 at 15 09 39" src="https://cloud.githubusercontent.com/assets/16718859/26746640/b8980b14-47a5-11e7-962f-ba8a872aa903.png">
<img width="133" alt="screen shot 2017-06-02 at 15 10 19" src="https://cloud.githubusercontent.com/assets/16718859/26746641/b8994c5e-47a5-11e7-80b4-416e4a895bc3.png">
<img width="148" alt="screen shot 2017-06-02 at 15 10 35" src="https://cloud.githubusercontent.com/assets/16718859/26746645/b8ad7648-47a5-11e7-8128-9164f694823c.png">

toger5 commented:
another (antialiased) version where of the editor theme where I try the rounded corners:
<img width="1680" alt="screen shot 2017-06-02 at 16 15 54" src="https://cloud.githubusercontent.com/assets/16718859/26748007/ddd64fea-47ae-11e7-91f7-84c764797260.png">

toger5 commented:
@volzhs @n-pigeon @djrm (pinging you: because of activity with UI related stuff) Would you mind to do some testing and think give your opinions on:
 - how corner radius and border width limiting should behave. (just leave artefacts when values are to extreme, decrease the radius... )
 - how border color should behave. (how far does the color go...)

akien-mga commented:
@toger5 Could you rebase and squash commits as you see best fit? I'm having nightmares of @reduz merging this without checking that there are 31 commits for a single feature ;)

n-pigeon commented:
@toger5
Wow, this is amazing! I will test when I will be available! This looks great! :)

n-pigeon commented:
@toger5
I've already see that your StyleBox is around 1px bigger than normal bounding boxes of Control? Check your Editor screenshot, where you have tab and tab window border your Stylebox is protruding from window by 1px.

toger5 commented:
@n-pigeon this is because the antialiasing is added around the style box... is I need to call rect.grow(-1) on the style when using aa.

toger5 commented:
@akien-mga rebased. what is your though? still too much?
 - adapted editor_themes.cpp
 - implemented grow(left,top,right,bottom) function
 - Activated draw_polygon function + renamings
I want those commits to be separate because they touch other parts not that relevant to the style box.
If there are still too much commits we can merge those:
	added expand margin + Complete renaming of things!  …			9c07937
 	added shadow  …			a5758bf
 	added anti aliasing  …			ec4f0b2
 	added corner detail property

akien-mga commented on scene/resources/style_box.cpp:
```diff
@@ -308,111 +352,590 @@ int StyleBoxFlat::_get_additional_border_size(Margin p_margin) const {
+	emit_changed();
+}
 float StyleBoxFlat::get_style_margin(Margin p_margin) const {
+	//	int margin_size = border_size;
```
Don't do that. Either drop the code, of if you *really* need to keep it around, use `#if 0 ... #endif` to disable it instead of commenting out all lines.

akien-mga commented on scene/resources/style_box.cpp:
```diff
@@ -308,111 +352,590 @@ int StyleBoxFlat::_get_additional_border_size(Margin p_margin) const {
+	rad = MIN(border_bottom, border_left);
+	inner_corner_radius[3] = MAX(corner_radius[3] - rad, 0);
+}
+inline void drawRing(Vector<Vector2> &verts, Vector<int> &indices, Vector<Color> &colors, const Rect2 style_rect, const int corner_radius[4],
```
camelCase...

akien-mga commented:
cf4777f was already merged in master, so it should be in this PR.

toger5 commented:
I just tried really hard to get rid of the commented code.
Didnt succeed yet...
I wanted to change the first commit where i introduce the comments. But than i have to merge the removed comments into every following commit..  And git is not doing a good job in evaluation merge conflict sections.
Is it okay to have one commit at the end which removes the comments?

Zireael07 commented:
Don't change the first commit, just make a commit that removes the comments and then rebase.

akien-mga commented:
> Don't change the first commit, just make a commit that removes the comments and then rebase.

Well my request *is* to change the first commit. But indeed it's non trivial rebasing work, as one needs to edit the first commit (easy), and then fix the merge conflicts when further commits are reapplied.

All in all, if tweaking the individual commits is too hard, you could also consider squashing everything together in one big commit (with detailed description of the features implemented) and then just amend the resulting commit to remove the commented code. The overall diff is not too bad: https://github.com/godotengine/godot/pull/8899/files (remember to remove cf4777f though!).

toger5 commented:
I might try the rebase again. Don't have a lot of time currently though... What is the timeline when I should be done? if I only have a little time I might just squash them all!

hubbyist commented:
I some times use a manual method to fix entangled multi commit branches, but before proceeding ask @akien-mga for approval that this approach is suitable for multi user repos as well.

Here is the method: 
Copy paste whole repo. Than on the original repo create an "archive/temp" branch on the same commit where the "entangled branch" currently is for safety (or get an third copy of repo). Create a new "corrected" branch to start from a suitable previous commit and checkout this branch. Then use a diff tool or copy paste changes by hand from "copy repo" to create new commits in this "corrected" branch. You can then reset the "entangled branch" to a appropriate commit and cherry pick commits from "corrected" branch and force push "entangled branch" which is not entangled anymore. 

Note: you may need to stash unrelated changes before starting as well.

toger5 commented:
@hubbyist this might actually be easier than what I tried...
I think I give the method:` commit with removed comments -> rebase -i to move it to the front -> squash and fix the merge issues ` another chance.

than I do do your approach

akien-mga commented:
Just noticed now that #8981 was not merged yet, so my suggestion to remove the `grow` commit was not doable... now it should be :)

toger5 committed:
NEW Style Box Flat Implementation
 - now use polygons!
 - renamed blend -> blend_border
 - draw_center -> filled
 - GDScript biding

toger5 committed:
Adapted godot to the new StyleBoxFlat

toger5 committed:
Added Corner Enum

toger5 committed:
new StyleBoxFlat added more drawing features
 - corner radius bindings
 - shadow
 - antiAliasing
 - CornerDetail

toger5 commented:
@akien-mga A complete new history ;)!

@hubbyist I did it similar to how you proposed it, Thank you!

@n-pigeon I also fixed the issue that anti-aliasing is making the style box slightly bigger.

The only thing that Is missing now is the border color.
Since this is completely changing how the editor UI is drawn it is important that we test properly. (performance and visual bugs... so any help is appreciated)

I don't know if I have time for the border color soon...

n-pigeon commented:
How you render gradient on borders? Still many polygons drawn with smaller shapes or you moved to shader? :)
If still polygons then it's a no, no, probably, someone that knows more about efficient rendering should take voice.

toger5 commented:
@n-pigeon I'm using vertex colors for colouring the style box, the way they work is that the vertex is coloured based on the three colors of the triangle corners with fading one into another.
This is the reason why the first OpenGL project ppl do is always the colourful triangle:
![](http://openglsamples.sourceforge.net/image/triangle.png)

This is extremely performant. "basically one of the simplest shader"
This is where I get my gradient from. so the perfomance is much better than the previous implementation when using blended borders. it's the same when drawing squared style boxes with corner_detail 1 and it is almost as performant as it can get when drawing more complex style boxes with rounded corners. Of course increasing the corner detail increases polygons -> makes it slower...

toger5 committed:
new StyleBoxFlat limits for drawing
 - corner and border are decreased if necassary to achoieve clean stylboxes
 - prohibits wired drawing artifacts when using wrong values
 - corner radius are relative to the partner corner when they would result in glitches

toger5 commented:
Another update:
AA now is really solid!
And the limitations make it hard to make wired style boxes.

n-pigeon commented:
@toger5 Good! Yeah I know how vertex shader works (wrote bunch of them xD)
Sound very good. Really great job. I will use your work a TON!

toger5 commented:
@akien-mga It does not compile because I use `INT32_MAX`
Am I supposed to use: `INT_MAX_MAX`
or: How can I get the max INT for the current system?

toger5 commented:
Finally the proper implementation of draw polygon is done. so I removed the commit where I implemented the _draw_polygon ( @akien-mga thanks for making me restructure the git history this made this task easier/possible ) and rebased.

toger5 commented:
The different border colors are a really hard task.. because I need to decide where I make the cut to a new color.. Also it makes it harder to change all the colors in the Inspector. you have to change each side...
So I might leave it out for now. If ppl really need multiple colors they still can use styleBoxTeture...

toger5 commented:
If so ill simplify it to one color and the pr would be done. I made some testing and it works really well.
definite good enough for the editor.

n-pigeon commented:
How UVs of border look like? If UV is done correctly, shader could be used to stylize them.

toger5 commented:
how would the UV's look like that you can make border dependent shaders?

toger5 commented:
currently there is no UV at all

Calinou commented:
Just saying, but I compiled this with MSVC 2017, replacing `INT32_MAX` with `INT_MAX`, and no styleboxes display at all in the editor (so it has a gray background and no visible buttons, but text is visible).

toger5 commented:
@Calinou I just forgot to remove them. thank you for the hint though. They get overwritten in the function directly afterwards anyways. I don't think they cause the issue that nothing gets drawn. That has something to do with how I use the new _draw_polygon function...
single color works. The issue with the poly only occur when passing an array longer than 1 to the color param of the function
I definitely need to remove the INT32_MAX to at least pass the checks.

n-pigeon commented:
@toger5 Tutorial time o/ :D 

They would be the same as frames themselves, just written into coordinates of square UV. So many styles would be possible...

![selection_130](https://user-images.githubusercontent.com/6152134/27109326-6ef7627e-50a2-11e7-9472-5d5478e946d5.png)

toger5 commented:
@n-pigeon so you want to use an additional texture for the colouring?
That definitely gives a lot of options, but than you also might just want to use a styleBoxTexture...
But It gives pretty clean results.. I'm thinking of generating those textures to use them for colouring. instead of using vertex colors... But that introduces another margin of error...
I think just making well designed options to allow for most designs might be best? thoughts?

option ideas:
border color for each border

corner style for each corner
options could be: fade, split, extend horizontally, extend vertically

your example 3 would have extend horizontally for : bottom left and bottom right
and example 2 and 4 would use split.

Zylann commented:
Don't forget you can still settle for a set of options that work well, and improve it later ;)

n-pigeon commented:
@toger5
It's not texture :D it's procedural shader no images. You can check .blend file :3rd_place_medal :3 
I only create coloring, shape would be from your polygon that you already are drawing. :)

Well most important part is for polygons to have usable UVs :)
Godot supports shaders for 2D, so with proper UVs on polygon user can create his own shaders and materials for shapes created with StyleBox.

Only question would be how much of styling implement into stylebox so it would be controlled from Theme resource.
I would support proper UVs even if you decide to only have one color for fill and border.
Your StyleBox will be probably used also in games, then proper UVs would give a lot of power to user.
For raw StyleBox controlled only by Theme I don't have opinion how much implement into stylebox.
Single color might be enough...

toger5 commented:
@n-pigeon sorry I didn't checkout the blend file. Got what you are doing now.
As soon as I get the new draw_poly function to work I will make an update to the pr with single border color. than we can discuss the addition of border color in another issue and make an update to the style box code. (Like @Zylann proposed)

n-pigeon commented:
Sure I agree with that :)
Proper UVs are just for to make this PR complete. Default UVs for polys in draw_polygon() is empty array which is a bit mueh :D

Zylann commented:
UVs can complexify fast as well, because you could not only need them on borders, but also in the middle, following a stretch or 9-slice approach, requiring more vertices again^^

n-pigeon commented:
@Zylann Not that many, it wouldn't by problem :)
I proposed UVs for border and fill on images above. Although your idea to use 9-slice maybe would be better (few more verts but this would be soooo little that it's no problem at all).

toger5 commented:
@akien-mga I did `commit --amend` `add` `push -f` ... to change to fix the current errors for windows + x11. But the PR does not update? the github branch does. How can I update?

karroffel commented:
@toger5 looks like it *did* update.

What's the state of it, is it complete or do you want to implement the border color stuff too?

toger5 commented:
@karroffel I just made a commit where I added an empty line, pushed, removed the commit pushed again with -f...

karroffel commented:
hmm okay, still doesn't answer the question if this is ready now or not from a feature standpoint :D

Also it builds just fine, so it looks like your old problem of the PR not updating was fixed.

toger5 commented:
@karroffel You are right missed your question.

It works pretty well right now. I'm still dabating about the multicolour border though. If I won't add it in this PR than I should remove the according function bindings from gdscript.
I also have a little performance improvement idea which I also want to add.

Ill definitely ping you guys as soon as I consider it ready to merge.
(sorry that this is taking so long. haven't got a lot of time for it recently...)

karroffel commented:
@toger5 awesome! We just had a PR discussion meeting on IRC and I wasn't quite sure what to do with this as it's not clear what the current state is.

Nice, just ping us when you're ready, no hurry :)

toger5 commented:
@karroffel sounds good! ;) feel free to mark it as WIP for now

toger5 committed:
new StyleBoxFlat removed multiBorderColor
 - removed only the bindings because the drawing code is not done yet
 - kept c++ functions for setting individual border color for future implementation

toger5 commented:
I just rebased again. And I think it is ready to merge. Editor works perfect with it.
There are still some edge cases where antiAliasing does not look well. But since the old style boxes also had issues with extreme values that might be fine. (at least it is definitely no step backwards)

toger5 commented:
@karroffel @akien-mga 
The pr got rebased on current master again.
Tested and works well.

karroffel commented:
Oh @toger5 there are conflicts now, can you resolve them? Then I'll merge :)

toger5 commented:
Nice. Cant do it right now bu i will tomorrow!

toger5 committed:
styleBoxFlat removed light, dark color entirely

toger5 commented:
Although there are no conflicts anymore don't merge yet.
want to add documentation and remove light, dark color.

akien-mga commented on scene/resources/theme.cpp:
```diff
@@ -929,8 +929,8 @@ RES ResourceFormatLoaderTheme::load(const String &p_path, const String &p_origin
-				sbflat->set_dark_color(dark);
-				sbflat->set_light_color(bright);
+				sbflat->set_border_color_all(bright);
+				//				sbflat->set_dark_color(dark);
```
If this is no longer relevant, best just remove it.

akien-mga commented on scene/resources/style_box.cpp:
```diff
@@ -285,156 +286,520 @@ void StyleBoxFlat::set_bg_color(const Color &p_color) {
+		}
+		//DRAW OUTER BORDER AA
+		if (!(border_width[0] == 0 && border_width[1] == 0 && border_width[2] == 0 && border_width[3] == 0)) {
+			//			for (int i = 0; i < 4; i++) {
```
Please remove commented out code.

akien-mga commented on scene/resources/style_box.cpp:
```diff
@@ -285,156 +286,520 @@ void StyleBoxFlat::set_bg_color(const Color &p_color) {
+			}
+		} else if (!(border_width[0] == 0 && border_width[1] == 0 && border_width[2] == 0 && border_width[3] == 0)) {
+			//DRAW INNER BORDER AA
+			//			for (int i = 0; i < 4; i++) {
```
Same here.

akien-mga commented on core/math/math_2d.h:
```diff
@@ -43,6 +43,14 @@ enum Margin {
 	MARGIN_BOTTOM
 };
 
+enum Corner {
```
If it's used only for StyleBoxFlat, should it really go in math_2d.h or could it be kept in style_box_flat.h?

toger5 replied:
I though it might could be used some where else so it does not hurt to give access to it by just importing math 2d...
But if you think it is improbable that it will be used seomwhere else I move it to style box

akien-mga commented:
I'd prefer to have the two documentation commits in a separate PR after this one is merged. If there were big issues in this one and we wanted to partially or totally revert it, the documentation changes would make it very difficult (also if anyone changes the documentation before this PR is merged, you might have to rebase).

akien-mga commented on scene/resources/style_box.cpp:
```diff
@@ -704,10 +704,11 @@ void StyleBoxFlat::_bind_methods() {
 
-	ClassDB::bind_method(D_METHOD("set_border_color", "margin", "color"), &StyleBoxFlat::set_border_color);
-	ClassDB::bind_method(D_METHOD("get_border_color", "margin"), &StyleBoxFlat::get_border_color);
+	//	ClassDB::bind_method(D_METHOD("set_border_color", "margin", "color"), &StyleBoxFlat::set_border_color);
```
Just get rid of comments, really.

akien-mga commented on scene/resources/style_box.cpp:
```diff
@@ -761,11 +762,11 @@ void StyleBoxFlat::_bind_methods() {
-	ADD_PROPERTYI(PropertyInfo(Variant::COLOR, "border_color_bottom"), "set_border_color", "get_border_color", MARGIN_BOTTOM);
+	ADD_GROUP("Border", "border_");
+	ADD_PROPERTY(PropertyInfo(Variant::COLOR, "border_color"), "set_border_color", "get_border_color");
+	//	ADD_PROPERTYI(PropertyInfo(Variant::COLOR, "border_color_top"), "set_border_color", "get_border_color", MARGIN_TOP);
```
Get rid of those.

akien-mga commented:
I still think there are too many commits for the overall changes. A rebase like this would be better IMO:
```
pick a129972	NEW Style Box Flat Implementation
pick 9585da8	Adapted godot to the new StyleBoxFlat	
pick 3fd82a9	Added Corner Enum
pick 8d2a439	new StyleBoxFlat added Corner Radius Bindings
squash 954d1fc	new StyleBoxFlat added Shadow
squash 192416b	new StyleBoxFlat added AntiAliasing
squash 90d8867	new StyleBoxFlat added Corner Detail
pick 2e0edae	new StyleBoxFlat limits for drawing
pick a08df42	new StyleBoxFlat removed multiBorderColor
squash 03681c1	styleBoxFlat removed light, dark color entirely
drop 5f252e8	Sync classes reference template with current code base
drop 177b1ef	documentation for StyleBoxFlat
```

toger5 commented:
@akien ty for thw review!
Also your rebase proposal makes sense!
Dont know if i will have internet in the next two weeks.(holiday)
Will push the changes afterwards

toger5 commented:
@akien-mga rebased + squashed + removed comments.
the only thing I didn't do yet, is changing the location of the Corner declaration.
I commented on your review. What are your thoughts.

kubecz3k commented:
from the meeting: @reduz approves but would like to hear also @karroffel and @akien-mga voice
